simulation/wallets.js

```diff
@@ -0,0 +1,516 @@
+    liveOperators[operatorIndex] = true
+    stakingOperators[operatorIndex] = true
+    operatorIndex++
+  }
+}
+
+// Create a new wallet every WALLET_CREATION_PERIOD days
+// FIXME: The actual wallet creation conditions are more complicated (min btc
+// and whatnot), so that's something to implement in the future.
+function createNewWalletEvent(day) {
+  if (day % WALLET_CREATION_PERIOD == 0) {
+    newWallet()
+  }
+}
+
+// Shuffle the staking operators and select 100 to form the signing group.
+function newWallet() {
+  walletBalances[walletIndex] = 0
+  const operators = getRandomSample(Object.keys(stakingOperators), WALLET_SIZE)
+  walletLiveOperators[walletIndex] = {}
+  walletStakingOperators[walletIndex] = {}
+  operators.forEach((operator) => {
+    walletLiveOperators[walletIndex][operator] = true
+    walletStakingOperators[walletIndex][operator] = true
+    let wallets = operatorToWallets[operator] || {}
+    wallets[walletIndex] = true
+    operatorToWallets[operator] = wallets
+  })
+  log(1, "creating new wallet index: " + walletIndex)
+
+  walletIndex++
+}
+
+// An implementation of wallet closure that transfers to a random live wallet.
+function randomTransferWithoutCap(wallet) {
+  let liveWallets = []
+  for (let i = 0; i < walletIndex; i++) {
+    if (
+      walletBalances[i] > WALLET_MIN_BTC &&
+      Object.keys(walletStakingOperators[i]).length >= HEARTBEAT &&
+      i != walletIndex
+    ) {
+      liveWallets.push(i)
+    }
+  }
+  const randomIndex = Math.floor(Math.random() * liveWallets.length)
+  const randomWallet = liveWallets[randomIndex]
+  log(
+    1,
+    "Transferring " +
+      walletBalances[wallet] +
+      " btc from Wallet#" +
+      wallet +
+      " to Wallet#" +
+      randomIndex
+  )
+  walletBalances[randomWallet] += walletBalances[wallet]
+  if (walletBalances[randomWallet] > biggestWalletBalance) {
+    biggestWalletBalance = walletBalances[randomWallet]
+  }
+  numberOfTransfers++
+}
+
+// An implementation of wallet closure that transfers to random live wallet(s)
+// sending out batches of `WALLET_TRANSFER_MAX` before picking a new wallet. If we
+// run out of wallets we start over.
+function randomTransfer(wallet) {
+  let liveWallets = []
+  for (let i = 0; i < walletIndex; i++) {
+    if (
+      walletBalances[i] > WALLET_MIN_BTC &&
+      Object.keys(walletStakingOperators[i]).length >= HEARTBEAT &&
+      i != walletIndex
+    ) {
+      liveWallets.push(i)
+    }
+  }
+  const transferCount = Math.ceil(walletBalances[wallet] / WALLET_TRANSFER_MAX)
+  const randomIndexes = getRandomSample(liveWallets, transferCount)
+  let remaining = walletBalances[wallet]
+  randomIndexes.forEach((randomIndex) => {
+    let transferAmount = 0
+    if (remaining > WALLET_TRANSFER_MAX) {
+      transferAmount = WALLET_TRANSFER_MAX
+    } else {
+      transferAmount = remaining
+    }
+    remaining -= transferAmount
+    const randomWallet = liveWallets[randomIndex]
```

Comment:
I don't think this is correct.
Le't say `walletBalances[]` looks this way and we try to close wallet#3:
```
key   walletBalances[key]
3     250
5     20
6     80
8     4
```
Then `liveWallets[]` may look this way:
```
key   liveWallets[key]
0     5
1     6
2     8
```
Then `randomIndexes[]` may look this way:
```
key   randomIndexes[key]
0     8
1     5
```
We are inside `randomIndexes.forEach((randomIndex)...` loop. We are executing the body of the loop for elements `randomIndex=8` and `randomIndex=5`. The `liveWallets[randomIndex]` expression is undefined for those values.
I think we should remove the `const randomWallet = liveWallets[randomIndex]` line and later in code increase the balance of `walletBalances[randomIndex]` instead of `walletBalances[randomWallet]`.

Reply:
Great catch holy moly

Reply:
Okay, I think we're working now:

Transferring 1 btc from Wallet#9 to Wallet#12
Transferring 1 btc from Wallet#21 to Wallet#23
Transferring 189 btc from Wallet#33 to Wallet#36
Transferring 109 btc from Wallet#50 to Wallet#56
Transferring 10 btc from Wallet#54 to Wallet#64
Transferring 88 btc from Wallet#55 to Wallet#63
Transferring 104 btc from Wallet#57 to Wallet#60
Transferring 195 btc from Wallet#61 to Wallet#69
Transferring 107.5 btc from Wallet#65 to Wallet#69
Transferring 107.5 btc from Wallet#65 to Wallet#68
Transferring 111.5 btc from Wallet#64 to Wallet#73
Transferring 111.5 btc from Wallet#64 to Wallet#70
Transferring 101 btc from Wallet#66 to Wallet#74
Transferring 101 btc from Wallet#66 to Wallet#63
Transferring 1 btc from Wallet#63 to Wallet#72
Transferring 104.5 btc from Wallet#72 to Wallet#73
Transferring 104.5 btc from Wallet#72 to Wallet#75
Transferring 103 btc from Wallet#71 to Wallet#76
Transferring 103 btc from Wallet#71 to Wallet#79
Transferring 67.5 btc from Wallet#70 to Wallet#73
Transferring 169.5 btc from Wallet#74 to Wallet#83
Transferring 169.5 btc from Wallet#74 to Wallet#77
Transferring 105.75 btc from Wallet#73 to Wallet#79
Transferring 105.75 btc from Wallet#73 to Wallet#78
Transferring 28.5 btc from Wallet#75 to Wallet#84
Transferring 128 btc from Wallet#76 to Wallet#81
Transferring 128 btc from Wallet#76 to Wallet#77
Transferring 159.375 btc from Wallet#78 to Wallet#87
Transferring 159.375 btc from Wallet#78 to Wallet#79
Transferring 189.04166666666666 btc from Wallet#79 to Wallet#82
Transferring 189.04166666666666 btc from Wallet#79 to Wallet#84
Transferring 189.04166666666666 btc from Wallet#79 to Wallet#80
Transferring 105.5 btc from Wallet#77 to Wallet#83
Transferring 167.5 btc from Wallet#81 to Wallet#83
Transferring 167.5 btc from Wallet#81 to Wallet#85
Transferring 161.625 btc from Wallet#83 to Wallet#80
Transferring 161.625 btc from Wallet#83 to Wallet#86
Transferring 161.625 btc from Wallet#83 to Wallet#90
Transferring 161.625 btc from Wallet#83 to Wallet#85
Transferring 40.66666666666663 btc from Wallet#80 to Wallet#85
Transferring 193.58333333333326 btc from Wallet#84 to Wallet#87
Transferring 106 btc from Wallet#88 to Wallet#86
Transferring 106 btc from Wallet#88 to Wallet#95
Transferring 146.5972222222222 btc from Wallet#85 to Wallet#90
Transferring 146.5972222222222 btc from Wallet#85 to Wallet#94
Transferring 146.5972222222222 btc from Wallet#85 to Wallet#91
Transferring 105 btc from Wallet#89 to Wallet#86
Transferring 105 btc from Wallet#89 to Wallet#87
Transferring 174.48958333333331 btc from Wallet#87 to Wallet#93
Transferring 174.48958333333331 btc from Wallet#87 to Wallet#90
Transferring 174.48958333333331 btc from Wallet#87 to Wallet#98
Transferring 174.48958333333331 btc from Wallet#87 to Wallet#92
Transferring 183.2986111111111 btc from Wallet#91 to Wallet#98
Transferring 183.2986111111111 btc from Wallet#91 to Wallet#101
Transferring 96.33680555555543 btc from Wallet#90 to Wallet#97
Transferring 168.24479166666666 btc from Wallet#92 to Wallet#102
Transferring 168.24479166666666 btc from Wallet#92 to Wallet#101
Transferring 138.24479166666666 btc from Wallet#93 to Wallet#103
Transferring 138.24479166666666 btc from Wallet#93 to Wallet#98

I also went ahead and implemented the proposed algorithm from https://github.com/keep-network/tbtc-v2/pull/127 where we split the funds evenly rather than the batches of `WALLET_TRANSFER_MAX`. It's more elegant, easier to understand, and leads to a more even distribution of funds so wins all around

Reply:
[ffa3b24](https://github.com/keep-network/tbtc-v2/pull/168/commits/ffa3b247b8b290ae3c1fc8e52b382a03d4abdc8a)

Reply:
> Added some logging

I can't see that logging you mentioned, but other than that things look good now :)